src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,415 @@
+        raise ValueError("Malformed response from LFS batch endpoint") from err
+
+
+def _upload_lfs_object(obj: LfsBatchObject, oid2addop: Dict[str, CommitOperationAdd]):
+    add_op = oid2addop.get(obj["oid"])
+    if add_op is None:
+        raise ValueError(f"Unknown OID: {obj['oid']}")
+    upload_info = add_op.upload_info()
+    upload_action = obj.get("actions", {}).get("upload", None)
+    if upload_action is None:
+        # The file was already uploaded
+        logger.debug(
+            f"Content of file {add_op.path_in_repo} is already present upstream"
+            " - skipping upload"
+        )
+        return
+    with add_op.fileobj() as fileobj:
+        lfs_upload(
+            fileobj=fileobj,
+            upload_action=upload_action,
+            upload_info=upload_info,
+        )
+        logger.debug(f"{add_op.path_in_repo}: Upload successful")
+
+
+def _preupload_payload(add_operation: CommitOperationAdd):
+    upload_info = add_operation.upload_info()
+    return {
+        "path": add_operation.path_in_repo,
+        "sample": base64.b64encode(upload_info.sample).decode("ascii"),
+        "size": upload_info.size,
+        "sha": upload_info.sha256.hex(),
+    }
+
+
+class PreUploadFileResponse(TypedDict):
+    path: str
+    uploadMode: UploadMode
+
+
+def fetch_upload_modes(
+    additions: Iterable[CommitOperationAdd],
+    repo_type: str,
+    repo_id: str,
+    token: str,
+    revision: str,
+    endpoint: Optional[str] = None,
+) -> List[Tuple[CommitOperationAdd, UploadMode]]:
+    """
+    Requests the HF Hub to determine wether each input file should be
+    uploaded as a regular git blob or as git LFS blob.
+
+    Args:
+        additions (``Iterable`` of :class:`CommitOperationAdd`):
+            Iterable of :class:`CommitOperationAdd` describing the files to
+            upload to the HF hub.
+        repo_type (``str``):
+            Type of the repo to upload to: `"model"`, `"dataset"` or `"space"`.
+        repo_id (``str``):
+            A namespace (user or an organization) and a repo name separated
+            by a `/`.
+        token (``str``):
+            An authentication token ( See https://huggingface.co/settings/tokens )
+        revision (``str``):
+            The git revision to upload the files to. Can be any valid git revision.
+
+    Returns:
+        list of 2-tuples, the first element being the add operation and
+        the second element the associated upload mode
+
+    Raises:
+        :class:`requests.HTTPError`:
+            If the HF Hub API returned an error
+    """
+    endpoint = endpoint if endpoint is not None else ENDPOINT
+    headers = {"authorization": f"Bearer {token}"} if token is not None else None
+    payload = {"files": [_preupload_payload(op) for op in additions]}
+
+    resp = requests.post(
+        f"{endpoint}/api/{repo_type}s/{repo_id}/preupload/{revision}",
+        json=payload,
+        headers=headers,
+    )
+    resp.raise_for_status()
```

Comment:
That's the way we do it in the website

In my opinion the proposed change is better because we don't have to edit this code if the logic ever changes

Reply:
This case is fine IMO, the head requests have only been an issue during mass-downloads of READMEs from the hub (for metadata fixes), which I can totally live with :)